Improve config options and header placement, add more init/per-frame callbacks

BlockBuilder57 opened the pull request:
Apologies that this is not multiple PRs, I never made one for the older commits and it's too late at this point.

This adds #ifndef directives around all of the backend config options, so compilers can override the existing config options without having to fork the library. Along those lines, helper functions (notably, the ones for InputHelper) were moved into the include directory. This allows the files to be overwritten somewhat, which allowed me to access the input from Xenomods without having to entirely change my version of the nn headers.

Additionally, the existing (post)init was split into pre and post, and a pre new frame callback has also been introduced. I've been having troubles loading new fonts from a file, and this change enables that to occur. (FWIW, The best way I've found is just adding the fonts before initializing the backend.) I _think_ I've updated all instances of imgui_xeno_init in the README/examples, but let me know if I missed something.

BlockBuilder57 committed:
Allow predefining of backend config options

BlockBuilder57 committed:
Move helper headers into include/

This allows the helpers to be included by other toolchains.

BlockBuilder57 committed:
Add callbacks for pre/post init and new frames

This allows finer control over various ImGui events, such as font management.

roccodev commented on include/imgui_xeno.h:
```diff
@@ -10,10 +10,12 @@
+ * @param renderCallback the function to execute during rendering
  */
-extern "C" void imgui_xeno_init(InitFunc init, ProcDrawFunc renderCallback);
+extern "C" void imgui_xeno_init(InitFunc preInitCallback, InitFunc postInitCallback, NewFrameFunc newFrameCallback, ProcDrawFunc renderCallback);
```
We can keep compatibility with the old API by making the new callbacks optional, by keeping `imgui_xeno_init` as it was, and adding new `imgui_xeno_set_on_...` functions. (or `add_on_...` if we allow multiple callbacks)

roccodev commented on README.md:
```diff
@@ -72,15 +75,17 @@ A working example can be found in the `examples/` directory.
 #[link(name = "imgui_xeno")] 
 extern "C" {
-    fn imgui_xeno_init(init: *const c_void, draw: *const c_void);
+    fn imgui_xeno_init(preInit: *const c_void, postInit: *const c_void, newFrame: *const c_void, draw: *const c_void);
```
These names (and their usages/references in Rust files) should be snake case, i.e. `pre_init` etc.

Same for the exported functions in the same file.

roccodev reviewed (changes requested):
I've left some notes wrt naming and API compatibility. Looks good otherwise.

More standard backends would let the user handle initialization instead of callbacks, but we're already using a non-standard approach to begin with.

BlockBuilder57 committed:
Don't set pre-init and new frame callbacks in init

These are technically optional, and as Rocco stated, it's best not to modify the API if it's not required.

Also fixes a lack of snake_case in Rust sources.

BlockBuilder57 commented:
Sorry for not getting to this until now, I've been incredibly busy. This should be good with what you've requested, let me know if there are any other changes to make.

roccodev commented on include/imgui_xeno.h:
```diff
@@ -27,6 +27,20 @@ extern "C" void imgui_xeno_init(InitFunc init, ProcDrawFunc renderCallback);
+extern "C" void imgui_xeno_add_on_pre_init(InitFunc preInitCallback);
+
+/**
+ * Optionally egisters a callback that is executed just before the start of every frame.
```
```suggestion
 * Optionally registers a callback that is executed just before the start of every frame.
```

BlockBuilder57 replied:
me when i egister

roccodev reviewed (changes requested):
One small typo in the otherwise excellent documentation, ready to merge after this.

Thanks again!

BlockBuilder57 commented:
Fixed, sorry about that!
